internal/vcs/git/blob.go

```
@@ -38,7 +38,54 @@ func ReadFile(ctx context.Context, repo gitserver.Repo, commit api.CommitID, nam
 	return b, nil
 }
 
+// GetFileReader returns an io.ReadCloser reading from the named file at commit.
+// The caller should always close the reader after use
+func GetFileReader(ctx context.Context, repo gitserver.Repo, commit api.CommitID, name string) (io.ReadCloser, error) {
+	if Mocks.GetFileReader != nil {
+		return Mocks.GetFileReader(commit, name)
+	}
+
+	span, ctx := opentracing.StartSpanFromContext(ctx, "Git: GetFileReader")
+	span.SetTag("Name", name)
+	defer span.Finish()
+
+	name = util.Rel(name)
+	br, err := newBlobReader(ctx, repo, commit, name)
+	if err != nil {
+		return nil, err
+	}
+	return br, nil
+}
+
 func readFileBytes(ctx context.Context, repo gitserver.Repo, commit api.CommitID, name string, maxBytes int64) ([]byte, error) {
+	br, err := newBlobReader(ctx, repo, commit, name)
+	if err != nil {
+		return nil, err
+	}
+	defer br.Close()
+
+	r := io.Reader(br)
+	if maxBytes > 0 {
+		r = io.LimitReader(r, maxBytes)
+	}
+	data, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, err
+	}
+	return data, nil
+}
+
+type blobReader struct {
+	ctx    context.Context
+	repo   gitserver.Repo
+	commit api.CommitID
+	name   string
+	cmd    *gitserver.Cmd
+	rc     io.ReadCloser
+}
+
+func newBlobReader(ctx context.Context, repo gitserver.Repo, commit api.CommitID, name string) (*blobReader, error) {
+	// NOTE: Do we really want to panic here? Surely returning an error is enough?
 	ensureAbsCommit(commit)
```

This makes me nervous as it can panic. We use it in various places in the package and wondering whether returning an error in most cases is a better opion

---

If you don't see a reason to panic, let's return an error instead!

---

Agreed we shouldn't panic in this case, and by looking at [refs to this function](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@25d7efca5ab6262652f0a9b896ae35c0c750da74/-/blob/internal/vcs/git/revisions.go#L33:6&tab=references), we probably want to make it return an error instead.

---

Done